src/Commands/core/ArchiveDumpCommands.php

```diff
@@ -0,0 +1,593 @@
+     *
+     * The following directories would be excluded from a file archive:
+     * - css
+     * - js
+     * - styles
+     * - php
+     *
+     * @command archive:dump
+     * @aliases ard
+     *
+     * @option code Archive codebase.
+     * @option files Archive Drupal files.
+     * @option db Archive database SQL dump.
+     * @option destination The full path and filename in which the archive should be stored. If omitted, it will be saved to the drush-backups directory.
+     * @option overwrite Overwrite destination file if exists.
+     * @option description Describe the archive contents.
+     * @option tags Add tags to the archive manifest. Delimit several by commas.
+     * @option overwrite Do not fail if the destination file exists; overwrite it instead. Default is --no-overwrite.
+     * @option generator The generator name to store in the MANIFEST.yml file. The default is "Drush archive-dump".
+     * @option generatorversion The generator version number to store in the MANIFEST file. The default is Drush version.
+     * @option exclude-code-paths Comma-separated list of paths to exclude from the code archive.
+     *
+     * @optionset_sql
+     * @optionset_table_selection
+     *
+     * @bootstrap max configuration
+     *
+     * @param array $options
+     *
+     * @return string
+     *
+     * @throws \Exception
+     */
+    public function dump(array $options = [
+        'code' => false,
+        'files' => false,
+        'db' => false,
+        'destination' => null,
+        'overwrite' => false,
+        'description' => null,
+        'tags' => null,
+        'generator' => null,
+        'generatorversion' => null,
+        'exclude-code-paths' => null,
+    ]): string
+    {
+        $this->prepareArchiveDir();
+
+        if (!$options['code'] && !$options['files'] && !$options['db']) {
+            $options['code'] = $options['files'] = $options['db'] = true;
+        }
+
+        $components = [];
+
+        if ($options['code']) {
+            $components[] = [
+                'name' => self::COMPONENT_CODE,
+                'path' => $this->getCodeComponentPath($options),
+            ];
+        }
+
+        if ($options['files']) {
+            $components[] = [
+                'name' => self::COMPONENT_FILES,
+                'path' => $this->getDrupalFilesComponentPath(),
+            ];
+        }
+
+        if ($options['db']) {
+            $components[] = [
+                'name' => self::COMPONENT_DATABASE,
+                'path' => $this->getDatabaseComponentPath($options),
+            ];
+        }
+
+        return $this->createArchiveFile($components, $options);
+    }
+
+    /**
+     * Creates a temporary directory for the archive.
+     *
+     * @throws \Exception
+     */
+    protected function prepareArchiveDir(): void
+    {
+        $this->filesystem = new Filesystem();
+        $this->archiveDir = FsUtils::tmpDir(self::ARCHIVES_DIR_NAME);
```

Comment:
@weitzman: What do you think about wrapping the call to `register_shutdown_path` in a general-purpose Drush tmpDir method, available to commands that need it? Seems better to me than making each command implement bespoke cleanup.